probe-rs/src/probe/jlink/mod.rs

```diff
@@ -1023,39 +1023,43 @@ fn bits_to_byte(bits: impl IntoIterator<Item = bool>) -> u32 {
     bit_val
 }
 
-pub(crate) fn list_jlink_devices() -> Result<impl Iterator<Item = DebugProbeInfo>, DebugProbeError>
-{
-    Ok(jaylink::scan_usb()?.map(|device_info| {
-        let vid = device_info.vid();
-        let pid = device_info.pid();
-        let (serial, product) = if let Ok(device) = device_info.open() {
-            let serial = device.serial_string();
-            let serial = if serial.is_empty() {
-                None
-            } else {
-                Some(serial.to_owned())
-            };
-            let product = device.product_string();
-            let product = if product.is_empty() {
-                None
-            } else {
-                Some(product.to_owned())
-            };
-            (serial, product)
-        } else {
-            (None, None)
-        };
-        DebugProbeInfo::new(
-            format!(
-                "J-Link{}",
-                product.map(|p| format!(" ({})", p)).unwrap_or_default()
-            ),
-            vid,
-            pid,
-            serial,
-            DebugProbeType::JLink,
-        )
-    }))
+pub(crate) fn list_jlink_devices() -> Vec<DebugProbeInfo> {
+    match jaylink::scan_usb() {
+        Ok(devices) => devices
+            .map(|device_info| {
+                let vid = device_info.vid();
+                let pid = device_info.pid();
+                let (serial, product) = if let Ok(device) = device_info.open() {
+                    let serial = device.serial_string();
+                    let serial = if serial.is_empty() {
+                        None
+                    } else {
+                        Some(serial.to_owned())
+                    };
+                    let product = device.product_string();
+                    let product = if product.is_empty() {
+                        None
+                    } else {
+                        Some(product.to_owned())
+                    };
+                    (serial, product)
+                } else {
+                    (None, None)
+                };
+                DebugProbeInfo::new(
+                    format!(
+                        "J-Link{}",
+                        product.map(|p| format!(" ({})", p)).unwrap_or_default()
+                    ),
+                    vid,
+                    pid,
+                    serial,
+                    DebugProbeType::JLink,
+                )
+            })
+            .collect(),
+        Err(_) => Vec::new(),
```

Comment:
I would appreciate if the error would somehow be logged I guess. Because otherwise the user gets not hints whatsoever what went wrong :)
What do you think? :)

Reply:
Good idea, I thought jaylink might do its own logging of that but it doesn't look like it does. I'll add a log... do you think warning or info?

Reply:
Actually I am not entirely sure how to do logging properly tbh. I feel like "error" could be suggesting something went terribly wrong. But actually, on a really bad error the application would panic. So maybe this is an error. If you believe this is too strong (I am uncertain too) I would do a warning I guess.

Reply:
If you're trying to use a jlink it's a bad error, but if you're using an stlink or daplink you don't really care at all and it would be annoying to have it error every time you use it.

However as far as I can tell the only error which could be returned here is because rusb failed to enumerate devices, which we'll almost certainly have already hit when enumerating daplink and stlink devices anyway, so maybe it doesn't matter?

Reply:
> If you're trying to use a jlink it's a bad error, but if you're using an stlink or daplink you don't really care at all and it would be annoying to have it error every time you use it.

Ok, I agree with that :)



> However as far as I can tell the only error which could be returned here is because rusb failed to enumerate devices, which we'll almost certainly have already hit when enumerating daplink and stlink devices anyway, so maybe it doesn't matter?

Well, that's true for St-Link, not CMSIS-DAP v1 tho, because it uses hidapi :) Or am I mistaken here?

Since the default log level is WARN, either, warning and error would be annoying, so I think only info would be left. But I think info is too weak.

And I feel stupid discussing this minor detail in length, but I guess this discussion is long overdue for the entire crate :D

Reply:
True, but I guess you'd get the rusb error in the log from daplink before it falls back to HID.

Perhaps we could set it to `info` to not spam people who aren't using jlink, but if no probes are detected overall we could print a more helpful error message (see also #357) which details how to turn up the log level, common troubleshooting, driver installation, permissions, etc? Maybe a link to a page on probe.rs.

Reply:
Sounds good to me!